### PR TITLE
ZX Compilation Pass

### PR DESCRIPTION
**Pull request by willsimmons1465**
Hack month finally gave us the time to plumb together the loose fragments of ZX code in tket and test it properly to give us an actual compilation pass that utilises the ZX data structure! General findings from testing were that it is 2-4x as fast as PyZX (orders of magnitude faster than FullPeepholeOptimise), though [quizx](https://github.com/Quantomatic/quizx) is anticipated to be even faster at the expense of the generality of the diagrams supported in tket. Our extraction does not include all optimisations made in PyZX's implementation, so is mostly outperformed wrt 2qb gates but sometimes gives a better outcome, probably from identifying a different flow.

Need to ensure we communicate to users that this pass will only work on purely unitary circuits (no classical bits, no created/discarded qubits) due to the limitations of the circuit extraction algorithm. Planning to add a section to the user manual on the ZX module soon.

### CODE REVIEW

**Review by aborgna-q — APPROVED** [review]

File: tket/src/Predicates/include/Predicates/PassLibrary.hpp [hunk 1]
```diff
@@ -127,4 +127,18 @@ const PassPtr &CnXPairwiseDecomposition();
  */
 const PassPtr &RemoveImplicitQubitPermutation();
 
+/**
+ * Attempt to optimise the circuit by simplifying in ZX calculus and extracting
+ * a circuit back out.
+ *
+ * Due to limitations in extraction, will not work if the circuit contains
+ * created or discarded qubits.
```
[the remark this answers is not in the document]
Comment (willsimmons1465) [reply tket/src/Predicates/include/Predicates/PassLibrary.hpp:135]: Good idea, that would cover it

File: tket/src/ZX/include/ZX/ZXDiagram.hpp [hunk 2]
```diff
@@ -201,6 +201,8 @@ class ZXDiagram {
   ZXVert add_vertex(ZXType type, QuantumType qtype = QuantumType::Quantum);
   ZXVert add_vertex(
       ZXType type, const Expr& param, QuantumType qtype = QuantumType::Quantum);
+  ZXVert add_clifford_vertex(
```
[the remark this answers is not in the document]
Comment (willsimmons1465) [reply tket/src/ZX/include/ZX/ZXDiagram.hpp:204]: Same with the continuous parameter version, the interpretation of the parameter is dependent on the type of the generator. Details of the interpretation is available in the header for the generators https://github.com/CQCL/tket/blob/0501f553bb4a01f0f88bff5a153fc6684819c379/tket/src/ZX/include/ZX/ZXGenerator.hpp#L67

File: pytket/tests/zx_diagram_test.py [hunk 3]
```diff
@@ -819,6 +820,120 @@ def test_XY_YZ_extraction() -> None:
...
+    assert diag.count_vertices(ZXType.Triangle) == 0
+    assert diag.count_vertices(ZXType.ZXBox) == 0
+
+    Rewrite.rebase_to_zx().apply(diag)
+    t2 = tensor_from_mixed_diagram(diag)
+    t2 = t2 * (tensor[-1, -1] / t2[-1, -1])
+    assert np.allclose(tensor, t2)
+
+
+def joint_normalise(a, b):
```
Comment (aborgna-q) [inline pytket/tests/zx_diagram_test.py:881]: The CI requires a type hint here.

**Review by aborgna-q — DISMISSED** [review]
LGTM

**Review by aborgna-q — DISMISSED** [review]

**Review by aborgna-q — APPROVED** [review]

**Pull request merged**